lib/io/pwm.cpp

```diff
@@ -0,0 +1,151 @@
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, percentage cannot be greater than 1.0");
+    return core::Result::kFailure;
+  }
+  if (duty_cycle < 0.0) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, percentage cannot be less than 0.0");
+    return core::Result::kFailure;
+  }
+  const auto time_active = static_cast<std::uint32_t>(duty_cycle * period_);
+  return setDutyCycleByTime(time_active);
+}
+
+core::Result Pwm::setDutyCycleByTime(const std::uint32_t time_active)
+{
+  if (duty_cycle_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to find PWM duty cycle file while setting duty cycle");
+    return core::Result::kFailure;
+  }
+  if (time_active > period_) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, active time cannot be greater than period");
+    return core::Result::kFailure;
+  }
+  const auto num_bytes_written = write(duty_cycle_file_, &time_active, sizeof(time_active));
+  if (num_bytes_written != sizeof(time_active)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to write to PWM duty cycle file");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully set PWM active time to %d", time_active);
+  return core::Result::kSuccess;
+}
+
+core::Result Pwm::setPeriod(const std::uint32_t period)
+{
+  if (period_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to find PWM period file while setting period");
+    return core::Result::kFailure;
+  }
+  period_                      = period;
+  const auto num_bytes_written = write(period_file_, &period_, sizeof(period_));
+  if (num_bytes_written != sizeof(period_)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to write to PWM period file");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully set PWM period to %d", period_);
+  return core::Result::kSuccess;
+}
+
+core::Result Pwm::setPolarity(const Polarity polarity)
+{
+  if (polarity_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to find PWM polarity file while setting polarity");
+    return core::Result::kFailure;
+  }
+  std::uint8_t polarity_value  = polarity == Polarity::kActiveHigh ? 0 : 1;
+  const auto num_bytes_written = write(polarity_file_, &polarity_value, sizeof(polarity_value));
+  if (num_bytes_written != sizeof(polarity_value)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to write to PWM polarity file");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully set PWM polarity to %d", polarity_value);
+  return core::Result::kSuccess;
+}
+
+core::Result Pwm::setMode(const Mode mode)
+{
+  if (enable_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to find PWM enable file while setting run mode");
+    return core::Result::kFailure;
+  }
+  std::uint8_t mode_value      = mode == Mode::kStop ? 0 : 1;
```

Comment:
`const`

also ternary operator is questionable; i don't think you gain much over just using a simple if/else or a cast from bool to int